packages/mui-material/src/Chip/Chip.js

```diff
@@ -156,141 +158,190 @@ const ChipRoot = styled('div', {
             ? `rgba(${theme.vars.palette.text.primaryChannel} / 0.4)`
             : alpha(theme.palette.text.primary, 0.4),
         },
-        ...(ownerState.size === 'small' && {
-          fontSize: 16,
-          marginRight: 4,
-          marginLeft: -4,
-        }),
-        ...(ownerState.color !== 'default' && {
-          color: theme.vars
-            ? `rgba(${theme.vars.palette[ownerState.color].contrastTextChannel} / 0.7)`
-            : alpha(theme.palette[ownerState.color].contrastText, 0.7),
-          '&:hover, &:active': {
-            color: (theme.vars || theme).palette[ownerState.color].contrastText,
-          },
-        }),
       },
-      ...(ownerState.size === 'small' && {
-        height: 24,
-      }),
-      ...(ownerState.color !== 'default' && {
-        backgroundColor: (theme.vars || theme).palette[ownerState.color].main,
-        color: (theme.vars || theme).palette[ownerState.color].contrastText,
-      }),
-      ...(ownerState.onDelete && {
-        [`&.${chipClasses.focusVisible}`]: {
-          backgroundColor: theme.vars
-            ? `rgba(${theme.vars.palette.action.selectedChannel} / calc(${theme.vars.palette.action.selectedOpacity} + ${theme.vars.palette.action.focusOpacity}))`
-            : alpha(
-                theme.palette.action.selected,
-                theme.palette.action.selectedOpacity + theme.palette.action.focusOpacity,
-              ),
+      variants: [
+        {
+          props: { size: 'small' },
+          style: {
+            height: 24,
+            [`& .${chipClasses.icon}`]: {
+              fontSize: 18,
+              marginLeft: 4,
+              marginRight: -4,
+            },
+            [`& .${chipClasses.deleteIcon}`]: {
+              fontSize: 16,
+              marginRight: 4,
+              marginLeft: -4,
+            },
+          },
         },
-      }),
-      ...(ownerState.onDelete &&
-        ownerState.color !== 'default' && {
-          [`&.${chipClasses.focusVisible}`]: {
-            backgroundColor: (theme.vars || theme).palette[ownerState.color].dark,
+        ...Object.entries(theme.palette)
+          .filter(([, value]) => value.main && value.contrastText)
+          .map(([color]) => {
+            return {
+              props: { color },
+              style: {
+                backgroundColor: (theme.vars || theme).palette[color].main,
+                color: (theme.vars || theme).palette[color].contrastText,
+                [`& .${chipClasses.deleteIcon}`]: {
+                  color: theme.vars
+                    ? `rgba(${theme.vars.palette[color].contrastTextChannel} / 0.7)`
+                    : alpha(theme.palette[color].contrastText, 0.7),
+                  '&:hover, &:active': {
+                    color: (theme.vars || theme).palette[color].contrastText,
+                  },
+                },
+                [`&.${chipClasses.focusVisible}.${chipClasses.deletable}`]: {
```

Comment:
This bumps specificity, there might be a way to do it with https://github.com/mui/material-ui/tree/master/packages/pigment-css-react#styling-based-on-runtime-values. Should I try? Any other options?

Using the props object is not an option AFAIK as

`props: { color, onDelete: what would go here? }`

Reply:
Ok, so we can't use callback, which is something many people may hit as a problem, if they need to use this in combination with dynamic check (callback) - cc @brijeshb42 we need to document this.

For this particular case, we can add an new ownerState, something like `deletable`. We can do the similar with the `clikable` instance.

Reply:
+1 to @mnajdova's suggestion. It'd be better if we can introduce a different `ownerState` because it won't work with emotion as of now and might break the component.
I also have my own reluctance to use it in library code even with Pigment CSS since it'll introduce inline styles which might not agree with CSP. It's most suited for user code I'd say.
If we can introduce a work around in our code itself to make the variants work and without any breaking changes for the user, we should go that way.

Reply:
+1 for adding a new field to `ownerState`. I did it with some components as well. We should not use a callback in Material UI because of non-emotion support and CSP issue.

Reply:
Is the `ownerState` object spread on the component's props? I didn't know that 😅 

With that, there's no need for extra `ownerState` properties; it simplifies the code and avoids specificity bumps, updated 🙌🏼

Reply:
Guess I missed this PR https://github.com/mui/material-ui/pull/41219 😂